From d65d658942e2c7dc5ce29865df100c3bf287b155 Mon Sep 17 00:00:00 2001
From: Dean Sanner <dsanner@us.ibm.com>
Date: Fri, 25 Jan 2019 10:24:36 -0600
Subject: [PATCH v1 9001/9001] Support direct load of POWERVM from BOOTKERNEL
 partition

 PowerVM doesn't fit in the PAYLOAD section of PNOR.  To
 enable a common PNOR layout instead put PowerVM into
 BOOTKERNEL partition, and if PHYP payload, load BOOTKERNEL
 instead of PAYLOAD PNOR partition

Change-Id: Id19f3ce708b537a19256c22aab3cc1b3e83d4ab1
---
 src/include/usr/pnor/pnor_const.H               |  1 +
 src/usr/isteps/istep20/call_host_load_payload.C | 10 ++++++++--
 src/usr/pnor/pnor_utils.C                       |  3 +++
 3 files changed, 12 insertions(+), 2 deletions(-)

diff --git a/src/include/usr/pnor/pnor_const.H b/src/include/usr/pnor/pnor_const.H
index d2515b7..9f8f8ea 100644
--- a/src/include/usr/pnor/pnor_const.H
+++ b/src/include/usr/pnor/pnor_const.H
@@ -76,6 +76,7 @@ enum SectionId
     CENTAUR_HW_IMG, /**< Centaur HCODE Reference image */
     HDAT,           /**< HDAT data */
     EECACHE,
+    BOOTKERNEL,
 #endif
     NUM_SECTIONS,   /**< Number of defined sections */
 
diff --git a/src/usr/isteps/istep20/call_host_load_payload.C b/src/usr/isteps/istep20/call_host_load_payload.C
index 6251eaf..808d87d 100644
--- a/src/usr/isteps/istep20/call_host_load_payload.C
+++ b/src/usr/isteps/istep20/call_host_load_payload.C
@@ -5,7 +5,7 @@
 /*                                                                        */
 /* OpenPOWER HostBoot Project                                             */
 /*                                                                        */
-/* Contributors Listed Below - COPYRIGHT 2016,2017                        */
+/* Contributors Listed Below - COPYRIGHT 2016,2019                        */
 /* [+] International Business Machines Corp.                              */
 /*                                                                        */
 /*                                                                        */
@@ -120,7 +120,13 @@ void* call_host_load_payload (void *io_pArgs)
         // Load payload data in PHYP mode or in Sapphire mode
         if(is_sapphire_load() || is_phyp_load())
         {
-            l_err = load_pnor_section( PNOR::PAYLOAD, payloadBase );
+            PNOR::SectionId l_secID = PNOR::PAYLOAD;
+            if (is_phyp_load())
+            {
+                l_secID = PNOR::BOOTKERNEL;
+            }
+
+            l_err = load_pnor_section( l_secID, payloadBase );
             if ( l_err )
             {
                 TRACFCOMP(ISTEPS_TRACE::g_trac_isteps_trace,
diff --git a/src/usr/pnor/pnor_utils.C b/src/usr/pnor/pnor_utils.C
index 68df913..e59bb71 100644
--- a/src/usr/pnor/pnor_utils.C
+++ b/src/usr/pnor/pnor_utils.C
@@ -369,6 +369,7 @@ bool PNOR::isEnforcedSecureSection(const uint32_t i_section)
                i_section == HB_DATA ||
                i_section == SBE_IPL ||
                i_section == PAYLOAD ||
+               i_section == BOOTKERNEL ||
                i_section == SBKT ||
                i_section == OCC ||
                i_section == HCODE ||
@@ -450,6 +451,8 @@ const char * PNOR::SectionIdToString( uint32_t i_secIdIndex )
         "CENHWIMG",    /**< PNOR::CENTAUR_HW_IMG : Centaur HCODE ref image   */
         "HDAT",        /**< PNOR::HDAT           : Hdat Data */
         "EECACHE",     /**< PNOR::EECACHE        : Cached data from various EEPROMs */
+        "BOOTKERNEL",  /**< PNOR::BOOTKERNEL     : petitboot for OPAL, other PHYP load */
+
 #endif
     };
 
-- 
1.8.2.2

